feat: support cacert in configMaps

kinolaev opened the pull request:
**Please provide a description of this PR:**

This PR allows to store a trust bundle for client certificate validation in a ConfigMap.
For example, istiod will try to find envoy's secret `kubernetes-gateway://default/server-tls-cacert` in:
1. ConfigMap `default/server-tls` (added by this PR)
2. Secret `default/server-tls-cacert`
3. Secret `default/server-tls`

ConfigMap is a more natural storage type for a trust bundle, because trust bundles don't contain any secret info.

This change will make Istio compatible with [trust-manager](https://cert-manager.io/docs/trust/trust-manager/) (from the developer of cert-manager), which automates the management of configMaps containing trust bundles.

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [x] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

kinolaev committed:
feat: support cacert in configMaps

Signed-off-by: Sergei Nikolaev <kinolaev@gmail.com>

istio-testing commented:
Hi @kinolaev. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/istio/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=istio%2Fistio).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository.
</details>

Stevenjin8 commented:
/ok-to-test

istio-testing commented:
@kinolaev: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | 2064557d0f95a705600bd7fd5fa4d12949dd1763 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/release-notes_istio/1932106974544007168) | true | `/test release-notes`
gencheck_istio | 2064557d0f95a705600bd7fd5fa4d12949dd1763 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/gencheck_istio/1932106962812538880) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn reviewed (changes requested):
This shouldn't happen implicitly. We will want to explicitly support `kind: ConfigMap` in the API and then explicitly configure it to read a ConfigMap instead of a Secret.

Note we already do this for BackendTLSPolicy so it should be fairly simple

kinolaev commented:
Hello @howardjohn , I'm agree with you, but I in that case we need to add some fields to Server struct or ServerTLSSettings  struct. And according to PR https://github.com/istio/istio/pull/46518#issuecomment-1700585874 you want to avoid this. That is why I've proposed this minimal change to support configmaps.

Gateway API already has a field for CA certificates - [frontendValidation](https://gateway-api.sigs.k8s.io/reference/spec/#frontendtlsvalidation). If you want an explicit way, I think supporting it, would be the best option. If you agree, I could try to make a PR to API repo first and then update this PR.

howardjohn commented:
Ah sorry I didn't realize this was for client verification.

Are you trying to use Gateway API or IstioGateway?

If GWAPI then [frontendValidation](https://gateway-api.sigs.k8s.io/reference/spec/#frontendtlsvalidation) is prob the way to go

jaellio reviewed (commented):
@kinolaev You'll need to run `make gen` in the project root to resolve the gen check failures.

Do we already support the ConfigMap credential type [here](https://github.com/istio/istio/pull/56549/files#diff-114fbb3b3e24c53a9a60fa4d55d2dd0ee8700fee267033854e1ef4701b6fb8dcL183)?

kinolaev commented:
@jaellio , yes, configMaps are supported in istiod through the `configmap://` scheme, but I haven't found a way to pass an URI with this scheme from a Gateway to istiod. Because GWAPI certificateRef through the [conversion into Istio Gateway ServerTLSSetting CredentialName](https://github.com/istio/istio/blob/1.26.1/pilot/pkg/config/kube/gateway/conversion.go#L1907) is passed to istiod in a form of two URIs: `kubernetes-gateway://namespace/secretFromCertificateRef` and [`kubernetes-gateway://namespace/secretFromCertificateRef-cacert`](https://github.com/istio/istio/blob/1.26.1/pilot/pkg/security/model/authentication.go#L235) and there is no separate field for a trust bundle URI with `configmap://` scheme.

keithmattix commented:
+1 to this not being implicit. I think we probably want to have an explicit kind reference in the api

kinolaev commented:
Thank you for your opinions! Based on you preference of making it explicit while avoiding extending Istio Gateway ServerTLSSettings I found another option.

What do you think about adding support for `-cacert` URIs like `configmap://namespace/configmapname-cacert` to [ServerTLSSettings.credentialsNames](https://istio.io/latest/docs/reference/config/networking/gateway/#ServerTLSSettings-credential_names)?

This would make an implementation of [GWAPI frontendValidation](https://gateway-api.sigs.k8s.io/reference/spec/#frontendtlsvalidation) very easy.

For example:
```yaml
apiVersion: gateway.networking.k8s.io/v1
kind: Gateway
metadata:
  namespace: default
  name: my-tls-gateway
spec:
  listeners:
  - protocol: HTTPS
    port: 443
    tls:
      mode: Terminate
      certificateRefs:
      - name: tls-cert
      frontendValidation:
        caCertificateRefs:
        - group: ""
          kind: ConfigMap
          namespace: default
          name: bundle
```
could be converted to 
```yaml
apiVersion: networking.istio.io/v1
kind: Gateway
metadata:
  namespace: default
  name: my-tls-gateway
spec:
  servers:
  - port: { number: 443, name: https, protocol: HTTPS }
    hosts: ["*"]
    tls:
      mode: MUTUAL
      credentialNames:
      - tls-cert
      - configmap://default/bundle-cacert
```
Explicit. No new API. Easy to implement.

I've already tested this approach locally and I plan to update this PR soon. But please don't wait for the update to express whether you like the proposal or not)

istio-testing commented:
@kinolaev: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_istio | 2064557d0f95a705600bd7fd5fa4d12949dd1763 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/gencheck_istio/1932106962812538880) | true | `/test gencheck`
release-notes_istio | dd73fb0c3a939130d21aff144ea6a41609977afa | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/release-notes_istio/1933811307887202304) | true | `/test release-notes`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kinolaev: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
release-notes_istio | dd73fb0c3a939130d21aff144ea6a41609977afa | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/release-notes_istio/1933811307887202304) | true | `/test release-notes`
gencheck_istio | dd73fb0c3a939130d21aff144ea6a41609977afa | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/gencheck_istio/1933811296147345408) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kinolaev commented:
If you mean `configmap://` URI format, then yes, it is defined [here](https://github.com/istio/istio/blob/1.26.1/pilot/pkg/model/credentials/resource.go#L37-L38) and it is used in [BackendTLSPolicy](https://gateway-api.sigs.k8s.io/reference/spec/#backendtlspolicyvalidation) implementation. And I've just found that a configmap from BackendTLSPolicy.spec.validation.caCertificateRefs is converted to a `configmap://` URI and saved to [ClientTLSSettings.credentialName](https://github.com/istio/istio/blob/1.26.1/pilot/pkg/config/kube/gateway/backend_policies.go#L199).

Cool, then in this PR I will support `configmap://` URIs pointing to CA certificates in [ServerTLSSettings.credentialsNames](https://istio.io/latest/docs/reference/config/networking/gateway/#ServerTLSSettings-credential_names).

jaellio commented:
> @howardjohn , @jaellio , @keithmattix could you please comment on the new proposal [#56549 (comment)](https://github.com/istio/istio/pull/56549#issuecomment-2970849148) ?
> 
> Btw, I'm thinking about splitting this PR into two: work on `configmap://` URI support in [ServerTLSSettings.credentialsNames](https://istio.io/latest/docs/reference/config/networking/gateway/#ServerTLSSettings-credential_names) in this PR and move Gateway API frontendValidation into another PR.

@kinolaev Thanks for your work on this - I like this approach and you're good to move forward with adding tests, comments and validation.

@keithmattix If we can get this change in before 1.27 feature freeze, should we update the istio gateway api docs for `CredentialName` to document support for specifying a ConfigMap for a cacert for client validation?

kinolaev committed:
chore: add client CA certificate with credential names test

Signed-off-by: Sergei Nikolaev <kinolaev@gmail.com>

kinolaev committed:
chore: remove FrontendValidation conversion

Signed-off-by: Sergei Nikolaev <kinolaev@gmail.com>

kinolaev commented:
Sorry for delay, had a busy month. I've added tests and removed GWAPI frontend validation from this PR.

I don't think I have the bandwidth for the docs. And I've found that `configmap://` uris are not mentioned in the docs at all. I guess it is a private API that were added to support BackendTLSPolicy. Maybe we don't need to make it public and document it now.

From my point of view this PR is complete and ready to be merged. 2 failed tests aren't related to the PR as I understand. Do I have to add a release note?

kinolaev commented:
/test unit-tests-arm64

kinolaev committed:
chore: add release note

Signed-off-by: Sergei Nikolaev <kinolaev@gmail.com>

kinolaev committed:
chore: add docs to the release note

Signed-off-by: Sergei Nikolaev <kinolaev@gmail.com>

kinolaev commented:
I've added docs https://github.com/istio/istio.io/pull/16719 and referenced it the release note.

@howardjohn , @jaellio , @keithmattix I'd be really happy to get a review from you and see this PR merged in 1.27)

howardjohn commented on pilot/pkg/security/model/authentication.go:
```diff
@@ -251,13 +251,24 @@ func ApplyCredentialSDSToServerCommonTLSContext(tlsContext *tls.CommonTlsContext
-			tlsContext.TlsCertificateSdsSecretConfigs[i] = ConstructSdsSecretConfigForCredential(name, credentialSocketExist, push)
+		tlsContext.TlsCertificateSdsSecretConfigs = []*tls.SdsSecretConfig{}
+		for _, name := range tlsOpts.CredentialNames {
+			if strings.HasSuffix(name, SdsCaSuffix) {
```
I am worried about this. First, let me say I think our existing API is quite problematic which is not your fault of course -- but something that is widely used and that we need to keep compatibility with its behavior.

Today, a user can put `credentialName: blah-cacert` and it will be treated as an opaque string. Now, the behavior is changed and we are implicitly deriving that they mean to use this as-is.

howardjohn replied:
overloading the string content like `cacert://` wouldn't work well since we already used (abused) that to have a meaning about the type.

It does seem like a new field `caCertCredentialName` may really be appropriate. WDYT @keithmattix

howardjohn reviewed (commented):
Sorry to get back to this so late :grimacing:

kinolaev committed:
feat: move cacert from credentialNames to caCertCredentialName

Signed-off-by: Sergei Nikolaev <kinolaev@gmail.com>

istio-testing commented:
@kinolaev: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-ambient-owned-cni_istio | 483da1a678ffff7d6a2683e60a1365e6361b7a72 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/integ-ambient-owned-cni_istio/1951035716850946048) | false | `/test integ-ambient-owned-cni`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kinolaev: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-ambient-owned-cni_istio | 483da1a678ffff7d6a2683e60a1365e6361b7a72 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/integ-ambient-owned-cni_istio/1951035716850946048) | false | `/test integ-ambient-owned-cni`
gencheck_istio | 483da1a678ffff7d6a2683e60a1365e6361b7a72 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/gencheck_istio/1951035716418932736) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kinolaev: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_istio | 483da1a678ffff7d6a2683e60a1365e6361b7a72 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/gencheck_istio/1951035716418932736) | true | `/test gencheck`
integ-ambient-owned-cni_istio | 3dc33c80e7cdd414f130eebc6c3dfb8a5e8d5e4a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/integ-ambient-owned-cni_istio/1951039145774157824) | false | `/test integ-ambient-owned-cni`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kinolaev: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
integ-ambient-owned-cni_istio | 3dc33c80e7cdd414f130eebc6c3dfb8a5e8d5e4a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/integ-ambient-owned-cni_istio/1951039145774157824) | false | `/test integ-ambient-owned-cni`
gencheck_istio | fe6c8debdf6459a4caa9b7a7ad7026b8f11df9b7 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/gencheck_istio/1951052486655610880) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@kinolaev: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_istio | fe6c8debdf6459a4caa9b7a7ad7026b8f11df9b7 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/gencheck_istio/1951052486655610880) | true | `/test gencheck`
integ-ambient-owned-cni_istio | 59b0ac68ff4619bb43dce9b6a8a9d130d3a597af | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/56549/integ-ambient-owned-cni_istio/1951054565495279616) | false | `/test integ-ambient-owned-cni`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

howardjohn commented:
integ-ambient-owned-cni_istio is not related to this, separate issue . will review the rest :+1:

howardjohn commented on pilot/pkg/security/model/authentication.go:
```diff
@@ -264,6 +264,9 @@ func ApplyCredentialSDSToServerCommonTLSContext(tlsContext *tls.CommonTlsContext
 			ConstructSdsSecretConfigForCredential(tlsOpts.CredentialName, credentialSocketExist, push),
 		}
 	}
+	if tlsOpts.CaCertCredentialName != "" {
```
nice, easier than I thought :slightly_smiling_face:

kinolaev commented:
Thanks, @howardjohn ! Is it possible to backport this PR to release-1.27 branch? Do I have to create a separate PR for it?
